### ISSUE TITLE
Inconsistent behavior is overwriting input parameter from value and ValueFrom.ValueFrom.ConfigMapKeyRef

### ISSUE THREAD

**Issue by apiwoni** [argoproj/argo-workflows#10182]
### Pre-requisites

- [X] I have double-checked my configuration
- [X] I can confirm the issues exists when I tested with `:latest`
- [ ] I'd like to contribute the fix myself (see [contributing guide](https://github.com/argoproj/argo-workflows/blob/master/docs/CONTRIBUTING.md))

### What happened/what you expected to happen?

inputs.parameters.value and inputs.parameters.valueFrom should have consistent behavior and be overwritten by argument from the caller. I'm able to overwrite inputs.parameters.value but not inputs.parameters.valueFrom.

Expected order of precedence per docs should be arguments first, then value or valueFrom and then default.

As coded, current behavior overwrites value with argument.Value and valueFrom with argument.valueFrom but when caller tries to overwrite with argument.Value and inputs.parameters.valueFrom is defined then it takes precedence over argument.Value. See [ProcessArgs](https://github.com/argoproj/argo-workflows/blob/3b2626ff900aff2424c086a51af5929fb0b2d7e5/workflow/common/util.go#L123)

```
if argParam != nil {
	if argParam.Value != nil {
                // inParam.Value gets overwritten then if inParam.ValueFrom != nil && inParam.ValueFrom.ConfigMapKeyRef != nil 
                // it will be overwritten with  inParam.ValueFrom
		inParam.Value = argParam.Value
	} else {
                // if this is true and set then (inParam.ValueFrom != nil && inParam.ValueFrom.ConfigMapKeyRef != nil) is false and  exception can be thrown!!!
		inParam.ValueFrom = argParam.ValueFrom
	}
}

if inParam.ValueFrom != nil && inParam.ValueFrom.ConfigMapKeyRef != nil {
       inParam.Value = wfv1.AnyStringPtr(cmValue)
}
else {
      if inParam.Value == nil {
              return nil, errors.Errorf(errors.CodeBadRequest, "inputs.parameters.%s was not supplied", inParam.Name)
	}
}
```

ProcessArgs is pretty bad coding and someone should take a serious look at it!

Simple solution is to process args first then values then defaults.

Here's efficient and, I think, also correct control flow which does not do ConfigMap when not needed:

```
func ProcessArgs(tmpl *wfv1.Template, args wfv1.ArgumentsProvider, globalParams, localParams Parameters, validateOnly bool, namespace string, configMapInformer cache.SharedIndexInformer) (*wfv1.Template, error) {
	newTmpl := tmpl.DeepCopy()
	for i, inParam := range newTmpl.Inputs.Parameters {

  // argument value from argument (if supplied)
  argParam := args.GetParameterByName(inParam.Name)
  if argParam != nil {
    if argParam.Value != nil {
      inParam.Value = argParam.Value
    } else {
      // setting inParam.ValueFrom was used previously to set inParam.Value and is no longer needed here
      //inParam.ValueFrom = argParam.ValueFrom
      inParam.Value = argParam.ValueFrom
    }
  }

  // if no argument supplied and no parameter value then try ValueFrom
  if inParam.Value == nil && inParam.ValueFrom != nil && inParam.ValueFrom.ConfigMapKeyRef != nil {
    if configMapInformer != nil {
      // SubstituteParams is called only at the end of this method. To support parametrization of the configmap
      // we need to perform a substitution here over the name and the key of the ConfigMapKeyRef.
      cmName, err := substituteConfigMapKeyRefParam(inParam.ValueFrom.ConfigMapKeyRef.Name, globalParams)
      if err != nil {
        log.WithError(err).Error("unable to substitute name for ConfigMapKeyRef")
        return nil, err
      }
      cmKey, err := substituteConfigMapKeyRefParam(inParam.ValueFrom.ConfigMapKeyRef.Key, globalParams)
      if err != nil {
        log.WithError(err).Error("unable to substitute key for ConfigMapKeyRef")
        return nil, err
      }

      cmValue, err := GetConfigMapValue(configMapInformer, namespace, cmName, cmKey)
      if err != nil {
        if inParam.ValueFrom.Default != nil && errors.IsCode(errors.CodeNotFound, err) {
          inParam.Value = inParam.ValueFrom.Default
        } else {
          return nil, errors.Errorf(errors.CodeBadRequest, "unable to retrieve inputs.parameters.%s from ConfigMap: %s", inParam.Name, err)
        }
      } else {
        inParam.Value = wfv1.AnyStringPtr(cmValue)
      }
    }
  } 
//  else {
//    if inParam.Value == nil {
//      return nil, errors.Errorf(errors.CodeBadRequest, "inputs.parameters.%s was not supplied", inParam.Name)
//    }
//  }

   // if no argument supplied, no parameter value and no value from config map then try default value
   if inParam.Value == nil && inParam.Default != nil {
       inParam.Value = inParam.Default
   }

    if inParam.Value == nil {
        return nil, errors.Errorf(errors.CodeBadRequest, "inputs.parameters.%s was not supplied", inParam.Name)
    }
    
    newTmpl.Inputs.Parameters[i] = inParam
}
  ...
  }
```

### Version

v3.4.4

### Paste a small workflow that reproduces the issue. We must be able to run the workflow; don't enter a workflows that uses private images.

```YAML
apiVersion: v1
kind: ConfigMap
metadata:
  name: config-properties
  labels:
    workflows.argoproj.io/configmap-type: Parameter
data:
  message1: "message1 from config map"
  message2: "message2 from config map"
  message3default: "default message3 from config map"
  message3overwrite: "overwrite message3 from config map from the caller"
---
apiVersion: argoproj.io/v1alpha1
kind: WorkflowTemplate
metadata:
  name: wf-template-main
spec:
  entrypoint: main
  templates:
    - name: main
      inputs:
        parameters:
          - name: message1
            valueFrom:
              configMapKeyRef:
                name: config-properties
                key: message1
          - name: message2
            value: "message2 from default for input parameter"
          - name: message3
            valueFrom:
              configMapKeyRef:
                name: config-properties
                key: message3default
      steps:
        - - name: first-step
            template: step
            arguments:
              parameters:
                - name: message
                  value: "{{inputs.parameters.message1}}"
        - - name: second-step
            template: step2
            arguments:
              parameters:
                - name: message
                  value: "{{inputs.parameters.message2}}"
        - - name: third-step
            template: step3
            arguments:
              parameters:
              - name: message
                value: "{{inputs.parameters.message3}}"
    - name: step
      inputs:
        parameters:
          - name: message
      container:
        image: docker/whalesay:latest
        command: [cowsay]
        args: ["{{inputs.parameters.message}}"]

    - name: step2
      inputs:
        parameters:
        - name: message
      container:
        image: docker/whalesay:latest
        command: [cowsay]
        args: ["{{inputs.parameters.message}}"]

    - name: step3
      inputs:
        parameters:
        - name: message
      container:
        image: docker/whalesay:latest
        command: [ cowsay ]
        args: [ "{{inputs.parameters.message}}" ]
---
apiVersion: argoproj.io/v1alpha1
kind: Workflow
metadata:
  name: wf-parameter-overwrite
spec:
  entrypoint: main
  arguments:
    parameters:
      - name: message1
        value: "configmap parameter overwrite with argument"
      - name: message2
        value: "value parameter overwrite with argument"
      - name: message3
        valueFrom:
          configMapKeyRef:
            name: config-properties
            key: message3overwrite
  workflowTemplateRef:
    name: wf-template-main
```

### Logs from the workflow controller

```text
< message1 from config map >

< value parameter overwrite with argument >

/ overwrite message3 from config map from \
\ the caller                              /
```

### Logs from in your workflow's wait container

```text
kubectl logs -n argo -c wait -l workflows.argoproj.io/workflow=${workflow},workflow.argoproj.io/phase!=Succeeded
```

**Comment by sarabala1979** [comment]
@apiwoni Do you like to submit the PR for this change?

**Comment by apiwoni** [comment]
@sarabala1979 I'm not really setup for golang development at the moment so it would take me some time but if there's no one else who can do quick PR with my changes then I can try to set up my environment for PR.

**Comment by mshatkhin23** [comment]
Hi! It seems like there is a PR that fixes this issue thats been approved but hasn't been merged yet. Any remaining blockers? (happy to help with getting it over the line!)

We just ran into this same issue on our team.

**Comment by terrytangyuan** [comment]
cc @LilTwo who's working on that PR

**Comment by LilTwo** [comment]
@terrytangyuan it's me, I was waiting for approval, just updated the branch to latest master

**Comment by mshatkhin23** [comment]
Thanks @terrytangyuan and @LilTwo !
It looks like auto-merge is enabled and all running checks have passed. Is it being blocked on the canceled `CI / E2E Tests (test-functional, v1.25.11-k3s1, minimal)`?

Also side note - pretty interested in how you all grouped the different e2e tests if there are any docs around that
